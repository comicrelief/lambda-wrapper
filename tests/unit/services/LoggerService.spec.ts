import Winston from 'winston';

import {
  Context,
  DependencyInjection,
  LoggerService,
} from '@/src';
import mockEvent from '@/tests/mocks/aws/event.json';

const mockContext = { invokedFunctionArn: 'my-function' } as Context;

const getLogger = (event = mockEvent, context = mockContext) => {
  const di = new DependencyInjection({ dependencies: { LoggerService } }, event, context);
  return new LoggerService(di);
};

describe('unit.services.LoggerService', () => {
  const context = { invokedFunctionArn: 'my-function' } as Context;

  const axiosResponses = {
    UNDEFINED: undefined,
    EMPTY: {},
    HTTP_417: {
      status: 417,
      data: { data: 1 },
      extra: 2,
    },
  };

  afterEach(() => jest.clearAllMocks());

  describe('logger', () => {
    it('should return a logger', () => {
      const logger = getLogger(undefined, context);
      expect(logger.logger.constructor.name).toEqual('DerivedLogger');
    });

    it('should not call `Winston.createLogger` twice', () => {
      const winston = Symbol('winston') as unknown as Winston.Logger;
      const logger = getLogger(undefined, context);

      jest.spyOn(Winston, 'createLogger').mockImplementation(() => winston);

      // use the getter several times
      expect(logger.logger).toEqual(winston);
      expect(logger.logger).toEqual(winston);
      expect(logger.logger).toEqual(winston);

      expect(Winston.createLogger).toHaveBeenCalledTimes(1);
    });
  });

  describe('error', () => {
    Object.entries(axiosResponses).forEach(([key, axiosResponse]) => {
      it(`Trims down the axios error: ${key}`, () => {
        const logger = getLogger();
        const log = jest.fn();
        const fakeLogger = { log } as unknown as Winston.Logger;

        jest.spyOn(logger, 'logger', 'get').mockReturnValue(fakeLogger);

        const error = {
          isAxiosError: true,
          raiseOnEpsagon: true,
          config: {
            url: 'http://localhost:9999',
            method: 'get',
          },
          extra: 1,
          response: axiosResponse,
          message: 'some-message',
        };

        logger.error(error);

        const loggerCall = log.mock.calls[0][2].error;

        expect(loggerCall).toMatchSnapshot();
        expect('extra' in loggerCall).toEqual(false);

        if (axiosResponse) {
          expect('extra' in loggerCall.response).toEqual(false);
        }
      });
    });
  });

  describe('info', () => {
    Object.entries(axiosResponses).forEach(([key, axiosResponse]) => {
      it(`Trims down the axios error: ${key}`, () => {
        const logger = getLogger();
        const log = jest.fn();
        const fakeLogger = { log } as unknown as Winston.Logger;

        jest.spyOn(logger, 'logger', 'get').mockReturnValue(fakeLogger);

        const error = {
          isAxiosError: true,
          raiseOnEpsagon: true,
          config: {
            url: 'http://localhost:9999',
            method: 'get',
          },
          extra: 1,
          response: axiosResponse,
          message: 'some-message',
        };

        logger.info(error);

        const loggerCall = log.mock.calls[0][1];

        expect(loggerCall).toMatchSnapshot();
        expect('extra' in loggerCall).toEqual(false);

        if (axiosResponse) {
          expect('extra' in loggerCall.response).toEqual(false);
        }
      });
    });
  });

  describe('warning', () => {
    let LOGGER_SOFT_WARNING: string | undefined;

    beforeAll(() => {
      LOGGER_SOFT_WARNING = process.env.LOGGER_SOFT_WARNING;
    });

    afterAll(() => {
      process.env.LOGGER_SOFT_WARNING = LOGGER_SOFT_WARNING;
    });

    ([
      ['', 'error'],
      ['some-value', 'error'],
      ['false', 'error'],
      ['0', 'error'],
      ['1', 'info'],
      ['true', 'info'],
    ] as [string, 'error' | 'info'][]).forEach(([loggerSoftWarning, func]) => {
      it(`uses 'this.logger.${func}' in ${loggerSoftWarning}`, () => {
        process.env.LOGGER_SOFT_WARNING = loggerSoftWarning;
        const logger = getLogger();

        jest.spyOn(logger, func).mockImplementation(() => { /* no-op */ });

        logger.warning({});

        expect(logger[func]).toHaveBeenCalledTimes(1);
      });
    });
  });

  describe('object', () => {
    ([
      'error',
      'warning',
      'info',
    ] as const).forEach((level) => {
      [
        null,
        'a string',
        { a: 1 },
        { a: { b: null }, c: 'a string' },
      ].forEach((object) => {
        it(`Logs a '${JSON.stringify(object)}' with level: '${level}'`, () => {
          const logger = getLogger();
          let calledArgs: any[] = [];
          const fakeLog = (...args: any[]) => { calledArgs = args; };

          jest.spyOn(logger, level).mockImplementation(fakeLog);

          logger.object('My action', object, level);

          expect(logger[level]).toHaveBeenCalledTimes(1);
          expect(calledArgs[0]).toMatchSnapshot();
        });
      });
    });
  });
});
