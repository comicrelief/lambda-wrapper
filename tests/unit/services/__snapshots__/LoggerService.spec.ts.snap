// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit.services.LoggerService error Trims down the axios error: EMPTY 1`] = `
{
  "config": {
    "method": "get",
    "url": "http://localhost:9999",
  },
  "message": "some-message",
  "response": {
    "data": undefined,
    "status": undefined,
  },
}
`;

exports[`unit.services.LoggerService error Trims down the axios error: HTTP_417 1`] = `
{
  "config": {
    "method": "get",
    "url": "http://localhost:9999",
  },
  "message": "some-message",
  "response": {
    "data": {
      "data": 1,
    },
    "status": 417,
  },
}
`;

exports[`unit.services.LoggerService error Trims down the axios error: UNDEFINED 1`] = `
{
  "config": {
    "method": "get",
    "url": "http://localhost:9999",
  },
  "message": "some-message",
}
`;

exports[`unit.services.LoggerService info Trims down the axios error: EMPTY 1`] = `
{
  "config": {
    "method": "get",
    "url": "http://localhost:9999",
  },
  "message": "some-message",
  "response": {
    "data": undefined,
    "status": undefined,
  },
}
`;

exports[`unit.services.LoggerService info Trims down the axios error: HTTP_417 1`] = `
{
  "config": {
    "method": "get",
    "url": "http://localhost:9999",
  },
  "message": "some-message",
  "response": {
    "data": {
      "data": 1,
    },
    "status": 417,
  },
}
`;

exports[`unit.services.LoggerService info Trims down the axios error: UNDEFINED 1`] = `
{
  "config": {
    "method": "get",
    "url": "http://localhost:9999",
  },
  "message": "some-message",
}
`;

exports[`unit.services.LoggerService object Logs a '"a string"' with level: 'error' 1`] = `"My action: '"a string"'"`;

exports[`unit.services.LoggerService object Logs a '"a string"' with level: 'info' 1`] = `"My action: '"a string"'"`;

exports[`unit.services.LoggerService object Logs a '"a string"' with level: 'warning' 1`] = `"My action: '"a string"'"`;

exports[`unit.services.LoggerService object Logs a '{"a":{"b":null},"c":"a string"}' with level: 'error' 1`] = `
"My action: '{
    "a": {
        "b": null
    },
    "c": "a string"
}'"
`;

exports[`unit.services.LoggerService object Logs a '{"a":{"b":null},"c":"a string"}' with level: 'info' 1`] = `
"My action: '{
    "a": {
        "b": null
    },
    "c": "a string"
}'"
`;

exports[`unit.services.LoggerService object Logs a '{"a":{"b":null},"c":"a string"}' with level: 'warning' 1`] = `
"My action: '{
    "a": {
        "b": null
    },
    "c": "a string"
}'"
`;

exports[`unit.services.LoggerService object Logs a '{"a":1}' with level: 'error' 1`] = `
"My action: '{
    "a": 1
}'"
`;

exports[`unit.services.LoggerService object Logs a '{"a":1}' with level: 'info' 1`] = `
"My action: '{
    "a": 1
}'"
`;

exports[`unit.services.LoggerService object Logs a '{"a":1}' with level: 'warning' 1`] = `
"My action: '{
    "a": 1
}'"
`;

exports[`unit.services.LoggerService object Logs a 'null' with level: 'error' 1`] = `"My action: 'null'"`;

exports[`unit.services.LoggerService object Logs a 'null' with level: 'info' 1`] = `"My action: 'null'"`;

exports[`unit.services.LoggerService object Logs a 'null' with level: 'warning' 1`] = `"My action: 'null'"`;
